### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by marat-asadurian**
## Integrating SafeDigit iPaaS into Shopify template app

### CODE REVIEW

File: app/safedigit/safedigit.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import type { NewTenant } from "./types/NewTenant";
+import type { Webhooks } from "./types/Webhoook";
+import { DeliveryMethod, type WebhookHandler } from "@shopify/shopify-api";
+import type { WebhookConfig } from "node_modules/@shopify/shopify-app-remix/build/ts/server/config-types";
```
Comment (letstri) [inline app/safedigit/safedigit.ts:3]: do you can import it without `node_modules`? or maybe just from `@shopify/app-remix`?

File: app/safedigit/safedigit.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+export class SafeDigit {
+  public static baseUrl = "https://api.test.safedigit.io/public/api";
```
Comment (letstri) [inline app/safedigit/safedigit.ts]: why only test?

File: app/safedigit/safedigit.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+  async createTenant(
+    clientId: string,
+    shop: string,
+    accessToken: string | undefined,
+  ) {
+    if (!accessToken) {
+      throw new Error("Access token is required");
```
Comment (letstri) [inline app/safedigit/safedigit.ts:10]: do shopify have some errors pattern? maybe createError or something like that

File: app/safedigit/safedigit.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    try {
+      const newTenant: NewTenant = {
+        tenantId: SafeDigit.getTenantId(clientId, shop),
+        name: shop,
+        configuration: {
+          shopifyToken: accessToken,
+          shopifyShopDomain: shop,
+        },
+      };
+      return await fetch(`${SafeDigit.baseUrl}/${clientId}/create-tenant`, {
```
Comment (letstri) [inline app/safedigit/safedigit.ts:21]: which HTTP client use shopify?

File: app/safedigit/safedigit.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    } catch (error) {
+      console.error("Error creating tenant", error);
+    }
+  }
+
+  static getTenantId(clientId: string, shop: string): string {
+    return `${clientId}@${shop}`;
+  }
+
+  static convertFetchHeadersToAxios(fetchHeaders: Headers) {
```
Comment (letstri) [inline app/safedigit/safedigit.ts]: why you didn't use axios?

File: app/safedigit/safedigit.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+          key,
+          {
+            deliveryMethod: DeliveryMethod.Http,
+            callbackUrl: SafeDigit.getCallbackUrl(clientId, value),
+          },
+        ];
+      }),
+    );
+
+    retVal.APP_UNINSTALLED = [
```
Comment (letstri) [inline app/safedigit/safedigit.ts:53]: incorrect types, you reassigning one type to another, don't do it

File: app/safedigit/safedigit.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      };
+      return await fetch(`${SafeDigit.baseUrl}/${clientId}/create-tenant`, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(newTenant),
+      });
+    } catch (error) {
+      console.error("Error creating tenant", error);
```
Comment (letstri) [inline app/safedigit/safedigit.ts]: throw

File: app/safedigit/safedigit.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+import { DeliveryMethod, type WebhookHandler } from "@shopify/shopify-api";
+import type { WebhookConfig } from "node_modules/@shopify/shopify-app-remix/build/ts/server/config-types";
+
+export class SafeDigit {
+  public static baseUrl = "https://api.test.safedigit.io/public/api";
+
+  async createTenant(
+    clientId: string,
+    shop: string,
+    accessToken: string | undefined,
```
Comment (letstri) [inline app/safedigit/safedigit.ts]: accessToken?: string

File: app/safedigit/safedigit.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    ];
+    return retVal;
+  }
+
+  private static getCallbackUrl(clientId: string, callbackPath: string) {
+    if (callbackPath.startsWith("http")) {
+      return callbackPath;
+    } else {
+      if (callbackPath.startsWith("/")) {
+        callbackPath = callbackPath.substring(1);
```
Comment (letstri) [inline app/safedigit/safedigit.ts]: do not reassign params, pls create new variable

File: app/safedigit/safedigit.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      }
+      return `${SafeDigit.baseUrl}/${clientId}/${callbackPath}`;
+    }
+  }
+}
+
+export const safeDigit = new SafeDigit();
```
Comment (letstri) [inline app/safedigit/safedigit.ts:75]: why do you export instance of the class if every method of this class is static?

File: app/safedigit/types/Webhoook.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+export type Webhooks = {
+  [key: string]: string;
```
Comment (letstri) [inline app/safedigit/types/Webhoook.ts]: useless type, use can use Record<string, string>

File: app/safedigit/safedigit.ts (hunk 1 above)
Comment (marat-asadurian) [reply app/safedigit/safedigit.ts:3]: I spent like an hour on it, but couldn't find a solution.

File: app/safedigit/safedigit.ts (hunk 4 above)
Comment (marat-asadurian) [reply app/safedigit/safedigit.ts:21]: No idea, but I didn't want to add packages, and this one worked straight out of the box.

File: app/safedigit/safedigit.ts (hunk 5 above)
Comment (marat-asadurian) [reply app/safedigit/safedigit.ts]: This is a leftover, I'll remove it

File: app/safedigit/safedigit.ts (hunk 4 above)
Comment (letstri) [reply app/safedigit/safedigit.ts:21]: I guess shopify uses axios, can you check it?

**Pull request merged**